### ISSUE TITLE
index_select uses only the first 32 entries of an index that is a row of a 2-D table

### ISSUE THREAD

**Issue by tdoublep** [torch-spyre/torch-spyre#3851]
### What happened?

`index_select` on dim 0 only uses the first 32 entries of the index when the index is a row of a 2-D int32 table (`table[0]`). The other 32 output rows come back wrong. 32 is one int32 stick (128-byte stick / 4 bytes).

`.contiguous()` makes no difference, because the row is already contiguous at storage offset 0 and so is the same view. `.clone()` fixes it in eager but not when compiled.

In eager, only row 0 is affected. `table[1]` and later rows are correct, presumably because a view at a non-zero storage offset gets materialised on its way to the device.

No `FallbackWarning` in either mode, so this really is running on device.

### What did you expect to happen?

All 64 rows gathered, as on CPU.

### How can we reproduce it?

fp16:

```python
import torch
import torch_spyre  # noqa: F401

NUM_BLOCKS, BLOCK_SIZE, NUM_KV_HEADS, HEAD_SIZE = 8, 64, 8, 128
NUM_SLOTS = NUM_BLOCKS * BLOCK_SIZE

src = torch.randn(NUM_SLOTS, NUM_KV_HEADS, HEAD_SIZE, dtype=torch.float16).to("spyre")

# Index table: row b holds the BLOCK_SIZE row ids of page b.
pages = torch.tensor([3, 7, 1, 5, 0, 2, 6, 4], dtype=torch.int32)
table_cpu = pages.unsqueeze(1) * BLOCK_SIZE + torch.arange(BLOCK_SIZE, dtype=torch.int32)
table = table_cpu.to("spyre")

# Reference: the same gather on the host, over the bytes the device holds.
expected = src.cpu().index_select(0, table_cpu[0])


def gather_view(t, tbl):
    return t.index_select(0, tbl[0])


def gather_contiguous(t, tbl):
    return t.index_select(0, tbl[0].contiguous())


def gather_clone(t, tbl):
    return t.index_select(0, tbl[0].clone())


for mode in ("eager", "compile"):
    for name, fn in (
        ("table[0]", gather_view),
        ("table[0].contiguous()", gather_contiguous),
        ("table[0].clone()", gather_clone),
    ):
        f = torch.compile(fn, dynamic=False, backend="inductor") if mode == "compile" else fn
        got = f(src, table).cpu()
        ok = int((got == expected).all(-1).all(-1).sum())
        print(f"{mode:8s} {name:23s} {ok:2d}/{BLOCK_SIZE} rows gathered correctly")
```

Output:

```
eager    table[0]                32/64 rows gathered correctly
eager    table[0].contiguous()   32/64 rows gathered correctly
eager    table[0].clone()        64/64 rows gathered correctly
compile  table[0]                32/64 rows gathered correctly
compile  table[0].contiguous()   32/64 rows gathered correctly
compile  table[0].clone()        32/64 rows gathered correctly
```

The reference is gathered from `src.cpu()` rather than from the tensor before the transfer, because the host-to-device fp16 transfer is not bit-exact (max abs diff 0.0039 here). Taking it from the bytes the device holds keeps the comparison exact.

### Any environmental details we need to know?

- torch-spyre `d83995da4dfabcee036bd17ad60a73124a789a0b`
- ibm-deeptools `2.0.0-0.main.1+2242.6a8d70c_280`, ibm-flex `2.0.0-0.main.1+474.94b4ae7_284`, ibm-senlib `2.0.0-0.main.1+235.7f44c00_198`

### Anything else we need to know?

- Both the index row and the source row have to span more than one stick. A 32-wide index row is fine, and so is a `[512, 64]` fp16 source, which is one 64-element fp16 stick per row. `[512, 128]` fails.
- Same shape of problem as #3826, where a view's size is ignored in favour of its base. That one is a slice write, this one is a gather index.

Found in spyre-inference paged attention: the per-page KV read indexes the cache with a row of the slot-index table, so half of every page came back wrong. Worked around with `.clone()` in torch-spyre/spyre-inference#551, which only holds because that kernel runs eager.

**Comment by ains-validator** [comment]
@validate please write a validation plan for this issue.

**Comment by ains-validator** [comment]
**Validator**

On it — working on branch `pocket/3851`.

```
oc logs -f job/pocket-torch-spyre-torch-spyre-3851-1787148126 -n hub
```

**Comment by ains-validator** [comment]
**Validator**

That turn failed — `oc logs job/pocket-torch-spyre-torch-spyre-3851-1787148126` has the detail. Branch `pocket/3851`
still has anything that was committed. Mention @validate again to retry, including
anything you said while it was running.

**Comment by ains-validator** [comment]
@validate please write a validation plan for this issue.

**Comment by ains-validator** [comment]
**Validator**

On it — working on branch `pocket/3851`.

```
oc logs -f job/pocket-torch-spyre-torch-spyre-3851-1787148966 -n hub
```

**Comment by ains-validator** [comment]
**Validator**

That turn failed — `oc logs job/pocket-torch-spyre-torch-spyre-3851-1787148966` has the detail. Branch `pocket/3851`
still has anything that was committed. Mention @validate again to retry, including
anything you said while it was running.

**Comment by ains-validator** [comment]
@validate please write a validation plan for this issue.

**Comment by ains-validator** [comment]
**Validator**

That turn failed — `oc logs job/pocket-torch-spyre-torch-spyre-3851-1787149876` has the detail. Branch `pocket/3851`
still has anything that was committed. Mention @validate again to retry, including
anything you said while it was running.

**Comment by ains-validator** [comment]
@validate please write a validation plan for this issue.

**Comment by ains-validator** [comment]
**Validator**

That turn failed — `oc logs job/pocket-torch-spyre-torch-spyre-3851-1787151070` has the detail. Branch `pocket/3851`
still has anything that was committed. Mention @validate again to retry, including
anything you said while it was running.

**Comment by ains-validator** [comment]
@validate please write a validation plan for this issue.

**Comment by ains-validator** [comment]
**Validator**

Validation plan for #3852 (the draft fix — one file, `enforce_indirect_access_layout.py`, no test yet). Full six-heading version is in commit `cdf326b` on `pocket/3851` (`validation/issue-3851-index-select-index-stick-adjacency.md`); I don't have push rights here.

**Population.** 1 PR (#3852). Sibling family that must *not* be read as closed by it: #3826, #3798, #3770, #2803, #2688 (7 of 10 open). Evidence baseline: 58 merged non-bot `fix` PRs touching `ops/eager.py`, `_inductor/lowering.py`, `_inductor/decompositions.py`, `csrc/spyre_mem.cpp`.

**Baseline.** Reproducer 32/64 on five of six variants, 64/64 only eager+`.clone()` (yours, at `d83995d` — I have no device here, so re-measure on `main` first). Coverage of this path today: 0 — 89 tests in `test_indirect_access_gather.py`, none indexes with a row of a 2-D table, nothing references the pass. 33/58 (56.9%) of comparable fix PRs ship a test; 0/58 have harness evidence recorded in the hub, so that part can only be read in the PR.

**Pass condition** (one named commit, command + output pasted):

1. 64/64 rows on all six reproducer variants, eager and compiled.
2. New test in `test_indirect_access_gather.py` sweeping index width {32,64,128,256} × table rows {2,8} × row {0,1,last} × source row width {64,128} × index dtype {int32,int64} × {eager,compile}; the already-working shapes from the issue still pass.
3. That test fails with the fix reverted — both runs quoted.
4. It asserts the index-layout property (adjacent dims ending at the stick) via `indirect_access_common`, not values alone, and shows op specs unchanged for an already-contiguous index (your "cost nothing" claim).
5. The 11 gating checks resolved for the changed path green (`regression` is the required one and the gather config carries that label, so no new config), plus one run at `SENCORES=32` — the gather suite is pinned to 1 core and #2699 was blocked for that gap.
6. `--runxfail` sweep: 0 unexplained XPASS.

**Blind spots.** Reproducer compares against `src.cpu()`, so it can't see a transfer-side error. Multi-core is a spot check, not coverage. `upstream_tests_beta/test_indexing.yaml` and `model_ops_tests` are push-to-main only. Nothing here verifies spyre-inference paged attention or that the `.clone()` workaround can be dropped.
